### PR TITLE
simplify across() selection

### LINKED ISSUE #5951: Differing results when aggregating lists in tibbles depending on .cols spec in across
The issue I'm facing are inconsistent results for rowwise mutation/summary with across(where()) based selection of list columns compared to equivalent .cols specs (to the best of my understanding).                                                                                  
With dplyr 1.0.5, the two mutations in the reprex have identical results.
With dplyr 1.0.6 (and 1.0.7), they don't.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

# The issue I'm facing is with rowwise mutation/summary and across(where()) for list columns.
# With dplyr 1.0.5, the following two mutations had identical results.
# With dplyr 1.0.6 (and 1.0.7), they don't.

tib <- tibble(a=list(1:2,3:4),b=list(5,6),c=list(NULL,NULL))
tib %>% rowwise %>% mutate(sum=across(everything(),sum))
#> # A tibble: 2 x 4
#> # Rowwise: 
#>   a         b         c      sum$a    $b    $c
#>   <list>    <list>    <list> <int> <dbl> <int>
#> 1 <int [2]> <dbl [1]> <NULL>     3     5     0
#> 2 <int [2]> <dbl [1]> <NULL>     7     6     0
tib %>% rowwise %>% mutate(sum=across(where(is.list),sum))
#> # A tibble: 2 x 4
#> # Rowwise: 
#>   a         b         c      sum$a    $b    $c
#>   <list>    <list>    <list> <int> <dbl> <int>
#> 1 <int [2]> <dbl [1]> <NULL>     3     5     0
#> 2 <int [2]> <dbl [1]> <NULL>    NA    NA    NA

# Inhibiting interpretation when creating the tibble apparently resolves the issue when the aggregate function is adjusted accordingly
# (for this example, but not for my code in which the lists result from data transformations).

tib <- tibble(a=I(list(1:2,3:4)),b=I(list(5,6)),c=I(list(NULL,NULL)))
tib %>% rowwise %>% mutate(sum=across(everything(),~ sum(unlist(.x))))
#> # A tibble: 2 x 4
#> # Rowwise: 
#>   a         b         c         sum$a    $b    $c
#>   <I<list>> <I<list>> <I<list>> <int> <dbl> <int>
#> 1 <int [2]> <dbl [1]> <NULL>        3     5     0
#> 2 <int [2]> <dbl [1]> <NULL>        7     6     0
tib %>% rowwise %>% mutate(sum=across(where(is.list),~ sum(unlist(.x))))
#> # A tibble: 2 x 4
#> # Rowwise: 
#>   a         b         c         sum$a    $b    $c
#>   <I<list>> <I<list>> <I<list>> <int> <dbl> <int>
#> 1 <int [2]> <dbl [1]> <NULL>        3     5     0
#> 2 <int [2]> <dbl [1]> <NULL>        7     6     0

# Unlisting by itself does not resolve the issue.

tib <- tibble(a=list(1:2,3:4),b=list(5,6),c=list(NULL,NULL))
tib %>% rowwise %>% mutate(sum=across(everything(),~ sum(unlist(.x))))
#> # A tibble: 2 x 4
#> # Rowwise: 
#>   a         b         c      sum$a    $b    $c
#>   <list>    <list>    <list> <int> <dbl> <int>
#> 1 <int [2]> <dbl [1]> <NULL>     3     5     0
#> 2 <int [2]> <dbl [1]> <NULL>     7     6     0
tib %>% rowwise %>% mutate(sum=across(where(is.list),~ sum(unlist(.x))))
#> # A tibble: 2 x 4
#> # Rowwise: 
#>   a         b         c      sum$a    $b    $c
#>   <list>    <list>    <list> <int> <dbl> <int>
#> 1 <int [2]> <dbl [1]> <NULL>     3     5     0
#> 2 <int [2]> <dbl [1]> <NULL>    NA    NA    NA
```

Session info:

``` r
sessionInfo()
#> R version 3.5.2 (2018-12-20)
#> Platform: x86_64-pc-linux-gnu (64-bit)
#> Running under: Debian GNU/Linux 10 (buster)
#> 
#> Matrix products: default
#> BLAS: /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
#> LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so
#> 
#> locale:
#>  [1] LC_CTYPE=en_US.utf8       LC_NUMERIC=C             
#>  [3] LC_TIME=en_US.utf8        LC_COLLATE=en_US.utf8    
#>  [5] LC_MONETARY=en_US.utf8    LC_MESSAGES=en_US.utf8   
#>  [7] LC_PAPER=en_US.utf8       LC_NAME=C                
#>  [9] LC_ADDRESS=C              LC_TELEPHONE=C           
#> [11] LC_MEASUREMENT=en_US.utf8 LC_IDENTIFICATION=C      
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] dplyr_1.0.6
#> 
#> loaded via a namespace (and not attached):
#>  [1] knitr_1.33        magrittr_2.0.1    tidyselect_1.1.1  R6_2.5.0         
#>  [5] rlang_0.4.11      fansi_0.5.0       stringr_1.4.0     styler_1.5.1     
#>  [9] highr_0.9         tools_3.5.2       xfun_0.24         utf8_1.2.1       
#> [13] cli_3.0.1         DBI_1.1.1         withr_2.4.2       htmltools_0.5.1.1
#> [17] ellipsis_0.3.2    assertthat_0.2.1  yaml_2.2.1        digest_0.6.27    
#> [21] tibble_3.1.2      lifecycle_1.0.0   crayon_1.4.1      purrr_0.3.4      
#> [25] vctrs_0.3.8       fs_1.5.0          glue_1.4.2        evaluate_0.14    
#> [29] rmarkdown_2.9     reprex_2.0.0      stringi_1.7.3     compiler_3.5.2   
#> [33] pillar_1.6.1      generics_0.1.0    backports_1.2.1   pkgconfig_2.0.3
```

### PR DESCRIPTION
**Pull request by romainfrancois**
The `DataMask` class gains a `$all_types` slot that maintains the types of the columns (initially the types of each column of the data, but then also includes the types of added columns (from a `summarise()` or a `mutate()`). 

This way `$across_cols()` no longer has to differentiate between columns that have been previously used or not, the selection via `tidyselect::eval_select()` operates on the named list of ptypes instead of a mixture of the column from the full data when not yet used and the current slice when already used. 

also: 

closes #5951

### CODE REVIEW

File: R/mutate.R [hunk 1]
```diff
@@ -344,13 +343,21 @@ mutate_cols <- function(.data, ..., caller_env) {
...
+
+          for (j in seq_along(types)) {
+            mask$add_one(types_names[j], chunks_extracted[[j]], ptype = types[[j]])
+          }
+
+          new_columns[types_names] <- result
         } else {
           # treat as a single output otherwise
           name <- quo_data$name_auto
+          mask$add_one(name = name, chunks = chunks, ptype = vec_ptype(result))
```
Comment (DavisVaughan) [inline R/mutate.R:358]: Small note: You could move `ptype <- vec_ptype(result)` outside the `if / else` statement since you use it in both conditions

**Review by DavisVaughan — COMMENTED** [review]
This seems like a pretty nice simplification!

I wasn't aware we had dropped usage of the cache in `across()` in favor of the expression expansion, but since we are doing that I think it makes sense to nuke the cache entirely.

File: R/data-mask.R [hunk 2]
```diff
@@ -221,6 +180,9 @@ DataMask <- R6Class("DataMask",
     # grows (and sometimes shrinks) as new columns are added/removed
     all_vars = character(),
 
+    # ptypes of all the variables
+    all_types = list(),
+
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/data-mask.R:185]: And actually, `current_vars` is `names(current_types)` so maybe we just have to maintain `current_vars`

**Review by lionel- — COMMENTED** [review]
This PR is doing two things right?

1. Removing the cache because it is very hard to get correct semantics.
2. Maintain a prototype of the current state of the data frame to support `across()` selections.

The problem with relying on the prototype is that tidyselect predicates are supposed to be applied on the whole data. It looks like this behaviour isn't covered in our unit tests so I've just added some in https://github.com/tidyverse/dplyr/commit/55add922.

So maybe use `current_data()` instead?

Since we are removing the cache, should we draw up a plan to move more thoroughly towards expansions of `across()` calls? That would require a review of the UI (deprecating argument-passing via `...`, then taking lists of functions via `...` when the deprecation period has elapsed).

**Comment by lionel-** [comment]
> So maybe use current_data() instead?

It will be hard to get consistent results with grouped data frames, especially in the non-expanded case. I think I would expect across selections to be computed on the whole data rather than slices, as is the case in the expanded case.

I'm starting to think that we should have designed `across()` more like a macro than a function call from the start, perhaps with a mechanism to add other macros for packages that need to wrap `across()`. Then we wouldn't have the unexpanded case to worry about.

**Comment by romainfrancois** [comment]
Reworked to accommodate for this test: 

```r
test_that("across() predicates operate on whole data", {
  df <- tibble(
    x = c(1, 1, 2),
    g = c(1, 1, 2)
  )

  out <- df %>%
    mutate(across(where(~ n_distinct(.x) > 1), ~ .x + 10))

  exp <- tibble(
    x = c(11, 11, 12),
    g = c(11, 11, 12)
  )

  expect_equal(out, exp)


  out <- df %>%
    group_by(g) %>%
    mutate(across(where(~ n_distinct(.x) > 1), ~ .x + 10))

  exp <- tibble(
    x = c(11, 11, 12),
    g = c(1, 1, 2)
  ) %>%
    group_by(g)

  expect_equal(out, exp)
})
```

`across()` now uses `$current_data` that represents the current state of the entire ungrouped data, i.e. the uncapped data when doing `mutate()` and the combined data when doing `summarise()`. 

There is an issue with `summarise()` for recycling (left a TODO in the code) that I need to think more about.

File: R/data-mask.R [hunk 3]
```diff
@@ -19,6 +19,7 @@ DataMask <- R6Class("DataMask",
       private$all_vars <- names_bindings
       private$data <- data
       private$caller <- caller
+      private$all_types <- vec_ptype(as.data.frame(data))
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/data-mask.R]: Eventually went with `private$current_data <- unclass(data)` we only need a list here.

**Pull request merged**